### PR TITLE
Parameterise ClusterIssuer for Dex, Gangway, HTTPBin

### PR DESCRIPTION
**Pull request by surajssd**
**Note:** See changes per commit.

* certmanager: Parameterise components to provide ClusterIssuer
    
    Add a variable `certmanager_cluster_issuer` which allow components to specify which cluster issuer to be used with Certmanager. This variable is added for components Dex, Gangway, HTTPBin.

* ci: Use certmanager clusterissuer `letsencrypt-staging`
    
    This will avoid the throttling done by LetsEncrypt when running in CI. Hence it will avoid the AWS e2e test failure.

* docs: Add variable `certmanager_cluster_issuer`
    
    Add docs to Dex, Gangway, Httpbin configuration reference guide.
    
Fixes https://github.com/kinvolk/lokomotive/issues/437

### CODE REVIEW

**Comment by surajssd** [comment]
So the tests failed with weird error:

```
    TestAWSIngress: aws_test.go:49: got an HTTP error: Get "https://httpbin.ci1590496091-tf.test.lokomotive-k8s.net/get": dial tcp 18.157.129.234:443: connect: connection refused
    TestAWSIngress: aws_test.go:49: got an HTTP error: Get "https://httpbin.ci1590496091-tf.test.lokomotive-k8s.net/get": x509: certificate signed by unknown authority
```

Are we hitting: https://github.com/golang/go/issues/24652 ?

**Comment by johananl** [comment]
> So the tests failed with weird error:
> 
> ```
>     TestAWSIngress: aws_test.go:49: got an HTTP error: Get "https://httpbin.ci1590496091-tf.test.lokomotive-k8s.net/get": dial tcp 18.157.129.234:443: connect: connection refused
>     TestAWSIngress: aws_test.go:49: got an HTTP error: Get "https://httpbin.ci1590496091-tf.test.lokomotive-k8s.net/get": x509: certificate signed by unknown authority
> ```
> 
> Are we hitting: [golang/go#24652](https://github.com/golang/go/issues/24652) ?

No, this is expected. We need to modify the test to ignore self-signed certs, or just test HTTP.

**Comment by surajssd** [comment]
> No, this is expected. We need to modify the test to ignore self-signed certs, or just test HTTP.

I can modify the test to accept self signed certs, but does that mean this PR is irrelevant?

**Comment by johananl** [comment]
> > No, this is expected. We need to modify the test to ignore self-signed certs, or just test HTTP.
> 
> I can modify the test to accept self signed certs, but does that mean this PR is irrelevant?

No, why irrelevant? We definitely want to use the staging environment assuming we're already randomizing the hostname and are hitting the maximum **certs per email** LE limit.

File: docs/configuration-reference/components/gangway.md [hunk 1]
```diff
@@ -80,6 +80,7 @@ Example:
 |------------------|-----------------------------------------------------------------------------------------------|:-------:|:--------:|
 | `cluster_name`   | The name of the cluster.                                                                      | -       | true     |
 | `ingress_host`   | Used as the `hosts` domain in the ingress resource for gangway that is automatically created. | -       | true     |
+| `certmanager_clusterissuer` | Cluster Issuer to be used by Certmanager while issuing TLS certificates. Supported values: `letsencrypt-production` & `letsencrypt-staging`. | letsencrypt-production | false |
```
Comment (johananl) [inline docs/configuration-reference/components/gangway.md:83]: Could you please fix the capitalization here?

Comment (johananl) [inline docs/configuration-reference/components/gangway.md:83]: I would replace the `&` with a comma.

File: pkg/components/dex/component.go [hunk 2]
```diff
@@ -245,19 +245,22 @@ type staticClient struct {
...
-	IssuerHost           string         `hcl:"issuer_host,attr"`
-	Connectors           []connector    `hcl:"connector,block"`
-	StaticClients        []staticClient `hcl:"static_client,block"`
-	GSuiteJSONConfigPath string         `hcl:"gsuite_json_config_path,optional"`
+	IngressHost              string         `hcl:"ingress_host,attr"`
+	IssuerHost               string         `hcl:"issuer_host,attr"`
+	Connectors               []connector    `hcl:"connector,block"`
+	StaticClients            []staticClient `hcl:"static_client,block"`
+	GSuiteJSONConfigPath     string         `hcl:"gsuite_json_config_path,optional"`
+	CertManagerClusterIssuer string         `hcl:"certmanager_clusterissuer,optional"`
```
Comment (ipochi) [inline pkg/components/dex/component.go:253]: Just a suggestion to use `certmanager_cluster_issuer`, in accordance with the annotation thats generally used ` cert-manager.io/cluster-issuer: nameOfClusterIssuer` ?

Comment (surajssd) [reply pkg/components/dex/component.go]: Done.

File: test/ingress/aws/aws_test.go [hunk 3]
```diff
@@ -43,8 +45,10 @@ func TestAWSIngress(t *testing.T) {
 
 	h := i.Spec.Rules[0].Host
 
+	httpsClient := getHTTPClient()
```
Comment (johananl) [inline test/ingress/aws/aws_test.go:48]: Nit: I'd call this var `client` or even `c` to follow Go's recommendations for var naming in local scope.

Comment (surajssd) [reply test/ingress/aws/aws_test.go:48]: done.

File: test/ingress/aws/aws_test.go [hunk 4]
```diff
@@ -67,3 +70,51 @@ func TestAWSIngress(t *testing.T) {
...
+VjrTSCLzO9Rv6s5wi03qoWvKAQQAElqTYRHhynJ3w6wuvKYF5zcZF3MDnrVGLbh1
+Q9ePRFBCiXOQ6wPLoUhrrbZ8LpFUFYDXHMtYM7P9sc9IAWoONXREJaO08zgFtMp4
+8iyIYUyQAbsvx8oD2M8kRvrIRSrRJSl6L957b4AFiLIQ/GgV2curs0jje7Edx34c
+idWw1VrejtwclobqNMVtG3EiPUIpJGpbMcJgbiLSmKkrvQtGng==
+-----END CERTIFICATE-----`
+
+	rootCAs := x509.NewCertPool()
+	if ok := rootCAs.AppendCertsFromPEM([]byte(letsEncryptStagingRootPEM)); !ok {
+		// This should fail in the developer testing itself.
+		panic("Failed to parse root certificate")
```
Comment (invidian) [inline test/ingress/aws/aws_test.go:110]: Using `t.Fatal()` would be better here, as `panic` will abort all the tests.

Comment (iaguis) [reply test/ingress/aws/aws_test.go:110]: I think panic is ok here. This function decodes a PEM-encoded certificate which is statically defined in code so it should not fail during regular testing.

Comment (invidian) [reply test/ingress/aws/aws_test.go:110]: Yeah, it's very nit-picking from my side :joy:

**Pull request merged**